eclair-core/src/main/scala/fr/acinq/eclair/transactions/Transactions.scala

```diff
@@ -386,62 +384,135 @@ object Transactions {
-   *   - [[ClaimHtlcTimeoutTx]] spends htlc-sent outputs of [[CommitTx]] after a timeout
+   * @param txIn        wallet input.
+   * @param spentOutput utxo spent by this wallet input.
+   */
+  case class WalletInput(txIn: TxIn, spentOutput: TxOut) {
+    val amount: Satoshi = spentOutput.amount
+  }
+
+  /**
+   * Whenever possible, [[ForceCloseTransaction]]s pay on-chain fees by lowering their output amount.
+   * When this isn't possible, we add wallet inputs to allow paying on-chain fees.
    *
-   * When *remote* *revoked* [[CommitTx]] is published:
-   *   - When using the default commitment format, [[ClaimP2WPKHOutputTx]] spends to-local output of [[CommitTx]]
-   *   - When using anchor outputs, [[ClaimRemoteDelayedOutputTx]] spends to-local output of [[CommitTx]]
-   *   - When using anchor outputs, [[ClaimAnchorOutputTx]] spends to-local anchor of [[CommitTx]]
-   *   - [[MainPenaltyTx]] spends remote main output using the per-commitment secret
-   *   - [[HtlcSuccessTx]] spends htlc-sent outputs of [[CommitTx]] for which they have the preimage (published by remote)
-   *     - [[ClaimHtlcDelayedOutputPenaltyTx]] spends [[HtlcSuccessTx]] using the revocation secret (published by local)
-   *   - [[HtlcTimeoutTx]] spends htlc-received outputs of [[CommitTx]] after a timeout (published by remote)
-   *     - [[ClaimHtlcDelayedOutputPenaltyTx]] spends [[HtlcTimeoutTx]] using the revocation secret (published by local)
-   *   - [[HtlcPenaltyTx]] spends competes with [[HtlcSuccessTx]] and [[HtlcTimeoutTx]] for the same outputs (published by local)
+   * @param inputs           inputs added by our bitcoin wallet.
+   * @param changeOutput_opt change output added by our bitcoin wallet, if any.
    */
+  case class WalletInputs(inputs: Seq[WalletInput], changeOutput_opt: Option[TxOut]) {
+    val amountIn: Satoshi = inputs.map(_.amount).sum
+    val fee: Satoshi = amountIn - changeOutput_opt.map(_.amount).getOrElse(0 sat)
+    val txIn: Seq[TxIn] = inputs.map(_.txIn)
+    val txOut: Seq[TxOut] = changeOutput_opt.toSeq
+    val spentUtxos: Map[OutPoint, TxOut] = inputs.map(i => i.txIn.outPoint -> i.spentOutput).toMap
+
+    /** Set the change output. */
+    def setChangeOutput(amount: Satoshi, changeScript: ByteVector): WalletInputs = {
+      val changeOutput = TxOut(amount, changeScript)
+      this.copy(changeOutput_opt = Some(changeOutput))
+    }
+
+    /** Set the change output amount. */
+    def setChangeAmount(amount: Satoshi): WalletInputs = {
+      this.copy(changeOutput_opt = changeOutput_opt.map(_.copy(amount = amount)))
+    }
+  }
+
+  /** Transactions spending a [[CommitTx]] or one of its descendants. */
   sealed trait ForceCloseTransaction extends TransactionWithInputInfo {
+    // @formatter:off
+    def commitmentFormat: CommitmentFormat
+    def expectedWeight: Int
+    // @formatter:on
+
+    def sign(): Transaction
```

Comment:
Why not have a `signedTx` property instead ?

Reply:
Signing is very fast and afaict we will always call `sign()` at least once so I don't think this is a problem.